zeppelin-zengine/src/main/java/org/apache/zeppelin/scheduler/pool/DynamicThreadPool.java

```diff
@@ -0,0 +1,273 @@
+        LOGGER.info("Have all the cron tasks been completed? {}", done);
+      } catch (InterruptedException e) {
+        Thread.currentThread().interrupt();
+      }
+    } else {
+      executor.shutdownNow();
+      final List<Runnable> rejected = executor.shutdownNow();
+      // Should be zero because of SynchronousQueue
+      LOGGER.info("Rejected tasks: {}: {}", rejected.size(), rejected);
+    }
+    INSTANCES.remove(schedulerInstanceId);
+  }
+
+  public void setThreadNamePrefix(String prfx) {
+    this.threadNamePrefix = prfx;
+  }
+
+  public String getThreadNamePrefix() {
+    return threadNamePrefix;
+  }
+
+  /**
+   * Set the number of worker threads in the pool. Increases <code>{@link
+   * ThreadPoolExecutor#maximumPoolSize}</code>
+   */
+  public void setThreadCount(int threadCount) throws SchedulerConfigException {
+    if (threadCount == this.threadCount) {
+      return;
+    }
+    LOGGER.info("Changing ThreadPool size from {} to {}", this.threadCount, threadCount);
+    if (threadCount <= 0) {
+      throw new SchedulerConfigException("Thread count must be > 0");
+    }
+    this.threadCount = threadCount;
+    executor.setMaximumPoolSize(threadCount);
+  }
+
+  public void setKeepAliveTime(long time, TimeUnit timeUnit) {
+    executor.setKeepAliveTime(time, timeUnit);
+  }
+
+  public void setMaximumPoolSize(int maximumPoolSize) throws SchedulerConfigException {
+    setThreadCount(maximumPoolSize);
+  }
+
+  public void setCorePoolSize(int corePoolSize) throws SchedulerConfigException {
+    setThreadCount(corePoolSize);
+  }
+
+  /**
+   * Get the maximum number of worker threads in the pool.
+   */
+  @Override
+  public int getPoolSize() {
+    return threadCount;
+  }
+
+  @Override
+  public void setInstanceId(String schedInstId) {
+    if (schedulerInstanceId != null && INSTANCES.get(schedulerInstanceId) != this) {
+      INSTANCES.remove(schedulerInstanceId);
+    }
+    this.schedulerInstanceId = schedInstId;
+    INSTANCES.put(schedulerInstanceId, this);
+  }
+
+  @Override
+  public void setInstanceName(String schedName) {
+    this.schedulerInstanceName = schedName;
+  }
+
+  /**
+   * Set the thread priority of worker threads in the pool.
+   */
+  public void setThreadPriority(int priority) {
+    this.priority = priority;
+  }
+
+  /**
+   * Get the thread priority of worker threads in the pool.
+   */
+  public int getThreadPriority() {
+    return priority;
+  }
+
+  private class DynamicThreadFactory implements ThreadFactory {
+
+    private final String nameFormat = threadNamePrefix + "-%d";
+    private final AtomicLong count = new AtomicLong(0L);
+
+    @Override
+    public Thread newThread(Runnable runnable) {
+      Thread newThread = new Thread(runnable);
+      newThread.setUncaughtExceptionHandler((th, e) -> LOGGER.error("ThreadPool error", e));
+      newThread.setName(String.format(nameFormat, count.getAndIncrement()));
```

Comment:
Мне кажется, лучше перестраховаться и сделать так:
```suggestion
      newThread.setName(String.format("%s-%d", threadNamePrefix, count.getAndIncrement()));
```
Иначе, если какой-нибудь умник укажет в конфиге `threadNamePrefix = %s` мы словим Exception. Ну и вообще, user input в шаблон String.format лучше не пускать :innocent: